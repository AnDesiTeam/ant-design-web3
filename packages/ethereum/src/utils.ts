import debug from 'debug';

const createDebug = (namespace: string) => {
  return debug(`antd-web3:ethereum:${namespace}`);
};

export { createDebug };
