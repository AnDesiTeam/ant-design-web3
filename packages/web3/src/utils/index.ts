export const getPlatform = () => {
  const userAgent = navigator.userAgent.toLowerCase();

  if (userAgent.includes('chrome')) {
    return 'Chrome';
  } else if (userAgent.includes('firefox')) {
    return 'Firefox';
  } else if (userAgent.includes('edge')) {
    return 'Edge';
  } else if (userAgent.includes('safari')) {
    return 'Safari';
  } else {
    return 'Other';
  }
};
