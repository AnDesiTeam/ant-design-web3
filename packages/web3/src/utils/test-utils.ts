import { vi } from 'vitest';

const oriClipboard = window.navigator.clipboard;
// 代理 navigator.clipboard 方法用于测试复制文本到粘贴板的功能
export const mockClipboard = () => {
  const clipboard = {
    text: '',
    writeText: (text: string) => {
      clipboard.text = text;
      return Promise.resolve();
    },
    readText: () => {
      return Promise.resolve(clipboard.text);
    },
  };
  Object.defineProperty(window, 'navigator', {
    value: {
      clipboard,
    },
  });

  return () => {
    Object.defineProperty(window, 'navigator', {
      value: {
        clipboard: oriClipboard,
      },
    });
  };
};

/**
 * Wait for a time delay. Will wait `advanceTime * times` ms.
 *
 * @param advanceTime Default 1000
 * @param times Default 20
 */
export async function waitFakeTimer(advanceTime = 1000, times = 20) {
  for (let i = 0; i < times; i += 1) {
    // eslint-disable-next-line no-await-in-loop
    await vi.waitFor(async () => {
      await Promise.resolve();

      if (advanceTime > 0) {
        vi.advanceTimersByTime(advanceTime);
      } else {
        vi.runAllTimers();
      }
    });
  }
}
