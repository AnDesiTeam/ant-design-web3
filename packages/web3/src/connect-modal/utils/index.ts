import type React from 'react';
import type { Wallet } from '../interface';
import { getPlatform } from '../../utils';

export const mergeReactNodeProps = (
  node: React.ReactNode,
  defaultNode: React.ReactNode,
  resolveString?: (node: string) => React.ReactNode,
) => {
  if (node === null || node === false) return node;
  if (typeof node === 'string' && resolveString) return resolveString(node);
  return node ?? defaultNode;
};

export const defaultGroupOrder = (a: string, b: string) => a.localeCompare(b);

export const getWalletRoute = (wallet: Wallet) => {
  const curExtensions =
    wallet?.extensions && wallet.extensions.some((item) => item.key === getPlatform());

  if (curExtensions) {
    return 'wallet';
  }

  if (wallet?.app) {
    return 'qrCode';
  }

  return 'unknown';
};
