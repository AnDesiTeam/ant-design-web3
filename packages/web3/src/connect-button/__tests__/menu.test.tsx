import { ConnectButton } from '..';
import { fireEvent, render } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { mockClipboard } from '../../utils/test-utils';
import { readCopyText } from '../../utils';
import type { MenuItemType } from 'antd/es/menu/hooks/useItems';

const menuItems: MenuItemType[] = [
  {
    label: 'Menu Item 1',
    key: '1',
    onClick: () => {
      console.log('Menu Item 1 clicked');
    },
  },
  {
    label: 'Menu Item 2',
    key: '2',
    onClick: () => {
      console.log('Menu Item 2 clicked');
    },
  },
];

describe('ConnectButton', () => {
  let resetMockClipboard: () => void;
  beforeEach(() => {
    resetMockClipboard = mockClipboard();
    vi.useFakeTimers();
  });
  afterEach(() => {
    resetMockClipboard();
    vi.useRealTimers();
  });
  it('Should show menu when hover button', async () => {
    const App = () => (
      <ConnectButton
        account={{
          address: '0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B',
        }}
        actionsMenu
      />
    );
    const { baseElement } = render(<App />);

    fireEvent.mouseEnter(baseElement.querySelector('.ant-dropdown-trigger') as Element);
    await vi.waitFor(() => {
      expect(baseElement.querySelector('.ant-dropdown')).not.toBeNull();
      expect(baseElement.querySelector('.ant-dropdown-menu')).not.toBeNull();
      expect(baseElement.querySelectorAll('.ant-dropdown-menu-item')?.length).toBe(2);
      expect(baseElement.querySelector('.ant-dropdown-menu-item')?.textContent).toBe(
        'Copy Address',
      );
      expect(
        baseElement.querySelector('.ant-dropdown-menu-item')?.getAttribute('data-menu-id'),
      ).toBe('rc-menu-uuid-test-copyAddress');
      expect(baseElement).toMatchSnapshot();
    });
    fireEvent.click(baseElement.querySelector('.ant-dropdown-menu-item') as Element);
    await vi.waitFor(() => {
      expect(readCopyText()).resolves.toBe('0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B');
      expect(baseElement.querySelector('.ant-message')).not.toBeNull();
      expect(baseElement.querySelector('.ant-message-notice-content')?.textContent).toBe(
        'Address Copied!',
      );
    });
  });
  it('Should show menu when hover button', async () => {
    const App = () => (
      <ConnectButton
        account={{
          address: '0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B',
        }}
        actionsMenu
      />
    );
    const { baseElement } = render(<App />);

    fireEvent.mouseEnter(baseElement.querySelector('.ant-dropdown-trigger') as Element);
    await vi.waitFor(async () => {
      expect(baseElement.querySelector('.ant-dropdown')).not.toBeNull();
    });
    fireEvent.click(baseElement.querySelector('.ant-web3-connect-button') as Element);
    await vi.waitFor(async () => {
      expect(baseElement.querySelector('.ant-dropdown')).toBeNull();
    });
  });
  it('Should not show menu by default', async () => {
    const App = () => <ConnectButton />;
    const { baseElement } = render(<App />);

    expect(baseElement.querySelector('.ant-dropdown-trigger')).toBeNull();
  });
  it('Should not show menu by actionsMenu be setting false', async () => {
    const App = () => <ConnectButton actionsMenu={false} />;
    const { baseElement } = render(<App />);

    expect(baseElement.querySelector('.ant-dropdown-trigger')).toBeNull();
  });
  it('Should insert menu items before default menu items when pass extraItems into actionsMenu', async () => {
    const menuClickFn = vi.fn();
    const App = () => (
      <ConnectButton
        account={{
          address: '0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B',
        }}
        actionsMenu={{
          extraItems: menuItems,
        }}
        onMenuClick={(info) => menuClickFn(info?.key)}
      />
    );
    const { baseElement } = render(<App />);

    fireEvent.mouseEnter(baseElement.querySelector('.ant-dropdown-trigger') as Element);
    await vi.waitFor(() => {
      expect(baseElement.querySelector('.ant-dropdown')).not.toBeNull();
      expect(baseElement.querySelector('.ant-dropdown-menu')).not.toBeNull();
      expect(baseElement.querySelectorAll('.ant-dropdown-menu-item')?.length).toBe(4);
      expect(baseElement.querySelector('.ant-dropdown-menu-item')?.textContent).toBe('Menu Item 1');
      expect(
        baseElement.querySelector('.ant-dropdown-menu-item')?.getAttribute('data-menu-id'),
      ).toBe('rc-menu-uuid-test-1');
      expect(baseElement).toMatchSnapshot();
    });
    fireEvent.click(baseElement.querySelectorAll('.ant-dropdown-menu-item')[0] as Element);
    await vi.waitFor(() => {
      expect(menuClickFn).toBeCalledWith('1');
    });
    fireEvent.click(baseElement.querySelectorAll('.ant-dropdown-menu-item')[2] as Element);
    await vi.waitFor(() => {
      expect(readCopyText()).resolves.toBe('0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B');
      expect(baseElement.querySelector('.ant-message')).not.toBeNull();
      expect(baseElement.querySelector('.ant-message-notice-content')?.textContent).toBe(
        'Address Copied!',
      );
      expect(menuClickFn).toBeCalledWith('copyAddress');
    });
  });
  it('Should override menu items when pass items into actionsMenu', async () => {
    const menuClickFn = vi.fn();
    const App = () => (
      <ConnectButton
        account={{
          address: '0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B',
        }}
        actionsMenu={{
          items: menuItems,
        }}
        onMenuClick={(info) => menuClickFn(info?.key)}
      />
    );
    const { baseElement } = render(<App />);

    fireEvent.mouseEnter(baseElement.querySelector('.ant-dropdown-trigger') as Element);
    await vi.waitFor(() => {
      expect(baseElement.querySelector('.ant-dropdown')).not.toBeNull();
      expect(baseElement.querySelector('.ant-dropdown-menu')).not.toBeNull();
      expect(baseElement.querySelectorAll('.ant-dropdown-menu-item')?.length).toBe(2);
      expect(baseElement.querySelector('.ant-dropdown-menu-item')?.textContent).toBe('Menu Item 1');
      expect(
        baseElement.querySelector('.ant-dropdown-menu-item')?.getAttribute('data-menu-id'),
      ).toBe('rc-menu-uuid-test-1');
      expect(baseElement).toMatchSnapshot();
    });
    fireEvent.click(baseElement.querySelectorAll('.ant-dropdown-menu-item')[0] as Element);
    await vi.waitFor(() => {
      expect(menuClickFn).toBeCalledWith('1');
    });
  });
});
