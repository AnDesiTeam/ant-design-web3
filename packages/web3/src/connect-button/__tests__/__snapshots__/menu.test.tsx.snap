// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConnectButton > Should insert menu items before default menu items when pass extraItems into actionsMenu 1`] = `
<body>
  <div>
    <button
      class="ant-btn css-dev-only-do-not-override-1ck3jst ant-btn-default ant-dropdown-trigger css-dev-only-do-not-override-1ck3jst ant-web3-connect-button ant-dropdown-open"
      type="button"
    >
      <div
        class="ant-web3-connect-button-text css-dev-only-do-not-override-1ck3jst"
      >
        Connect Wallet
      </div>
    </button>
  </div>
  <div>
    <div
      class="ant-dropdown ant-slide-up-appear ant-slide-up-appear-prepare ant-slide-up css-dev-only-do-not-override-1ck3jst ant-dropdown-placement-bottomLeft"
      style="--arrow-x: 0px; --arrow-y: 0px; left: -1000vw; top: -1000vh; box-sizing: border-box;"
    >
      <ul
        class="ant-dropdown-menu ant-dropdown-menu-root ant-dropdown-menu-vertical ant-dropdown-menu-light css-dev-only-do-not-override-1ck3jst"
        data-menu-list="true"
        role="menu"
        tabindex="0"
      >
        <li
          class="ant-dropdown-menu-item ant-dropdown-menu-item-only-child"
          data-menu-id="rc-menu-uuid-test-1"
          role="menuitem"
          tabindex="-1"
        >
          <span
            class="ant-dropdown-menu-title-content"
          >
            Menu Item 1
          </span>
        </li>
        <li
          class="ant-dropdown-menu-item ant-dropdown-menu-item-only-child"
          data-menu-id="rc-menu-uuid-test-2"
          role="menuitem"
          tabindex="-1"
        >
          <span
            class="ant-dropdown-menu-title-content"
          >
            Menu Item 2
          </span>
        </li>
        <li
          class="ant-dropdown-menu-item"
          data-menu-id="rc-menu-uuid-test-copyAddress"
          role="menuitem"
          tabindex="-1"
        >
          <span
            aria-label="copy"
            class="anticon anticon-copy ant-dropdown-menu-item-icon"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="copy"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M832 64H296c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h496v688c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8V96c0-17.7-14.3-32-32-32zM704 192H192c-17.7 0-32 14.3-32 32v530.7c0 8.5 3.4 16.6 9.4 22.6l173.3 173.3c2.2 2.2 4.7 4 7.4 5.5v1.9h4.2c3.5 1.3 7.2 2 11 2H704c17.7 0 32-14.3 32-32V224c0-17.7-14.3-32-32-32zM350 856.2L263.9 770H350v86.2zM664 888H414V746c0-22.1-17.9-40-40-40H232V264h432v624z"
              />
            </svg>
          </span>
          <span
            class="ant-dropdown-menu-title-content"
          >
            Copy Address
          </span>
        </li>
        <li
          class="ant-dropdown-menu-item"
          data-menu-id="rc-menu-uuid-test-disconnect"
          role="menuitem"
          tabindex="-1"
        >
          <span
            aria-label="login"
            class="anticon anticon-login ant-dropdown-menu-item-icon"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="login"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <defs>
                <style />
              </defs>
              <path
                d="M521.7 82c-152.5-.4-286.7 78.5-363.4 197.7-3.4 5.3.4 12.3 6.7 12.3h70.3c4.8 0 9.3-2.1 12.3-5.8 7-8.5 14.5-16.7 22.4-24.5 32.6-32.5 70.5-58.1 112.7-75.9 43.6-18.4 90-27.8 137.9-27.8 47.9 0 94.3 9.3 137.9 27.8 42.2 17.8 80.1 43.4 112.7 75.9 32.6 32.5 58.1 70.4 76 112.5C865.7 417.8 875 464.1 875 512c0 47.9-9.4 94.2-27.8 137.8-17.8 42.1-43.4 80-76 112.5s-70.5 58.1-112.7 75.9A352.8 352.8 0 01520.6 866c-47.9 0-94.3-9.4-137.9-27.8A353.84 353.84 0 01270 762.3c-7.9-7.9-15.3-16.1-22.4-24.5-3-3.7-7.6-5.8-12.3-5.8H165c-6.3 0-10.2 7-6.7 12.3C234.9 863.2 368.5 942 520.6 942c236.2 0 428-190.1 430.4-425.6C953.4 277.1 761.3 82.6 521.7 82zM395.02 624v-76h-314c-4.4 0-8-3.6-8-8v-56c0-4.4 3.6-8 8-8h314v-76c0-6.7 7.8-10.5 13-6.3l141.9 112a8 8 0 010 12.6l-141.9 112c-5.2 4.1-13 .4-13-6.3z"
              />
            </svg>
          </span>
          <span
            class="ant-dropdown-menu-title-content"
          >
            Disconnect
          </span>
        </li>
      </ul>
      <div
        aria-hidden="true"
        style="display: none;"
      />
    </div>
  </div>
</body>
`;

exports[`ConnectButton > Should insert menu items before default menu items when pass extraItems into actionsMenu 2`] = `
<body>
  <div>
    <button
      class="ant-btn css-dev-only-do-not-override-1ck3jst ant-btn-default ant-dropdown-trigger css-dev-only-do-not-override-1ck3jst ant-web3-connect-button ant-dropdown-open"
      type="button"
    >
      <div
        class="ant-web3-connect-button-text css-dev-only-do-not-override-1ck3jst"
      >
        <div
          class="ant-space css-dev-only-do-not-override-1ck3jst ant-space-horizontal ant-space-align-center ant-space-gap-row-small ant-space-gap-col-small ant-web3-address css-dev-only-do-not-override-1ck3jst"
        >
          <div
            class="ant-space-item"
          >
            <span>
              0x21CD...Fd3B
            </span>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div>
    <div
      class="ant-dropdown ant-slide-up-appear ant-slide-up-appear-start ant-slide-up css-dev-only-do-not-override-1ck3jst ant-dropdown-placement-bottomLeft"
      style="--arrow-x: 0px; --arrow-y: 0px; left: -1000vw; top: -1000vh; box-sizing: border-box; transition: none;"
    >
      <ul
        class="ant-dropdown-menu ant-dropdown-menu-root ant-dropdown-menu-vertical ant-dropdown-menu-light css-dev-only-do-not-override-1ck3jst"
        data-menu-list="true"
        role="menu"
        tabindex="0"
      >
        <li
          class="ant-dropdown-menu-item ant-dropdown-menu-item-only-child"
          data-menu-id="rc-menu-uuid-test-1"
          role="menuitem"
          tabindex="-1"
        >
          <span
            class="ant-dropdown-menu-title-content"
          >
            Menu Item 1
          </span>
        </li>
        <li
          class="ant-dropdown-menu-item ant-dropdown-menu-item-only-child"
          data-menu-id="rc-menu-uuid-test-2"
          role="menuitem"
          tabindex="-1"
        >
          <span
            class="ant-dropdown-menu-title-content"
          >
            Menu Item 2
          </span>
        </li>
        <li
          class="ant-dropdown-menu-item"
          data-menu-id="rc-menu-uuid-test-copyAddress"
          role="menuitem"
          tabindex="-1"
        >
          <span
            aria-label="copy"
            class="anticon anticon-copy ant-dropdown-menu-item-icon"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="copy"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M832 64H296c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h496v688c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8V96c0-17.7-14.3-32-32-32zM704 192H192c-17.7 0-32 14.3-32 32v530.7c0 8.5 3.4 16.6 9.4 22.6l173.3 173.3c2.2 2.2 4.7 4 7.4 5.5v1.9h4.2c3.5 1.3 7.2 2 11 2H704c17.7 0 32-14.3 32-32V224c0-17.7-14.3-32-32-32zM350 856.2L263.9 770H350v86.2zM664 888H414V746c0-22.1-17.9-40-40-40H232V264h432v624z"
              />
            </svg>
          </span>
          <span
            class="ant-dropdown-menu-title-content"
          >
            Copy Address
          </span>
        </li>
        <li
          class="ant-dropdown-menu-item"
          data-menu-id="rc-menu-uuid-test-disconnect"
          role="menuitem"
          tabindex="-1"
        >
          <span
            aria-label="login"
            class="anticon anticon-login ant-dropdown-menu-item-icon"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="login"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <defs>
                <style />
              </defs>
              <path
                d="M521.7 82c-152.5-.4-286.7 78.5-363.4 197.7-3.4 5.3.4 12.3 6.7 12.3h70.3c4.8 0 9.3-2.1 12.3-5.8 7-8.5 14.5-16.7 22.4-24.5 32.6-32.5 70.5-58.1 112.7-75.9 43.6-18.4 90-27.8 137.9-27.8 47.9 0 94.3 9.3 137.9 27.8 42.2 17.8 80.1 43.4 112.7 75.9 32.6 32.5 58.1 70.4 76 112.5C865.7 417.8 875 464.1 875 512c0 47.9-9.4 94.2-27.8 137.8-17.8 42.1-43.4 80-76 112.5s-70.5 58.1-112.7 75.9A352.8 352.8 0 01520.6 866c-47.9 0-94.3-9.4-137.9-27.8A353.84 353.84 0 01270 762.3c-7.9-7.9-15.3-16.1-22.4-24.5-3-3.7-7.6-5.8-12.3-5.8H165c-6.3 0-10.2 7-6.7 12.3C234.9 863.2 368.5 942 520.6 942c236.2 0 428-190.1 430.4-425.6C953.4 277.1 761.3 82.6 521.7 82zM395.02 624v-76h-314c-4.4 0-8-3.6-8-8v-56c0-4.4 3.6-8 8-8h314v-76c0-6.7 7.8-10.5 13-6.3l141.9 112a8 8 0 010 12.6l-141.9 112c-5.2 4.1-13 .4-13-6.3z"
              />
            </svg>
          </span>
          <span
            class="ant-dropdown-menu-title-content"
          >
            Disconnect
          </span>
        </li>
      </ul>
      <div
        aria-hidden="true"
        style="display: none;"
      />
    </div>
  </div>
</body>
`;

exports[`ConnectButton > Should override menu items when pass items into actionsMenu 1`] = `
<body>
  <div>
    <button
      class="ant-btn css-dev-only-do-not-override-1ck3jst ant-btn-default ant-dropdown-trigger css-dev-only-do-not-override-1ck3jst ant-web3-connect-button ant-dropdown-open"
      type="button"
    >
      <div
        class="ant-web3-connect-button-text css-dev-only-do-not-override-1ck3jst"
      >
        Connect Wallet
      </div>
    </button>
  </div>
  <div>
    <div
      class="ant-dropdown ant-slide-up-appear ant-slide-up-appear-prepare ant-slide-up css-dev-only-do-not-override-1ck3jst ant-dropdown-placement-bottomLeft"
      style="--arrow-x: 0px; --arrow-y: 0px; left: -1000vw; top: -1000vh; box-sizing: border-box;"
    >
      <ul
        class="ant-dropdown-menu ant-dropdown-menu-root ant-dropdown-menu-vertical ant-dropdown-menu-light css-dev-only-do-not-override-1ck3jst"
        data-menu-list="true"
        role="menu"
        tabindex="0"
      >
        <li
          class="ant-dropdown-menu-item ant-dropdown-menu-item-only-child"
          data-menu-id="rc-menu-uuid-test-1"
          role="menuitem"
          tabindex="-1"
        >
          <span
            class="ant-dropdown-menu-title-content"
          >
            Menu Item 1
          </span>
        </li>
        <li
          class="ant-dropdown-menu-item ant-dropdown-menu-item-only-child"
          data-menu-id="rc-menu-uuid-test-2"
          role="menuitem"
          tabindex="-1"
        >
          <span
            class="ant-dropdown-menu-title-content"
          >
            Menu Item 2
          </span>
        </li>
      </ul>
      <div
        aria-hidden="true"
        style="display: none;"
      />
    </div>
  </div>
</body>
`;

exports[`ConnectButton > Should override menu items when pass items into actionsMenu 2`] = `
<body>
  <div>
    <button
      class="ant-btn css-dev-only-do-not-override-1ck3jst ant-btn-default ant-dropdown-trigger css-dev-only-do-not-override-1ck3jst ant-web3-connect-button ant-dropdown-open"
      type="button"
    >
      <div
        class="ant-web3-connect-button-text css-dev-only-do-not-override-1ck3jst"
      >
        <div
          class="ant-space css-dev-only-do-not-override-1ck3jst ant-space-horizontal ant-space-align-center ant-space-gap-row-small ant-space-gap-col-small ant-web3-address css-dev-only-do-not-override-1ck3jst"
        >
          <div
            class="ant-space-item"
          >
            <span>
              0x21CD...Fd3B
            </span>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div>
    <div
      class="ant-dropdown ant-slide-up-appear ant-slide-up-appear-start ant-slide-up css-dev-only-do-not-override-1ck3jst ant-dropdown-placement-bottomLeft"
      style="--arrow-x: 0px; --arrow-y: 0px; left: -1000vw; top: -1000vh; box-sizing: border-box; transition: none;"
    >
      <ul
        class="ant-dropdown-menu ant-dropdown-menu-root ant-dropdown-menu-vertical ant-dropdown-menu-light css-dev-only-do-not-override-1ck3jst"
        data-menu-list="true"
        role="menu"
        tabindex="0"
      >
        <li
          class="ant-dropdown-menu-item ant-dropdown-menu-item-only-child"
          data-menu-id="rc-menu-uuid-test-1"
          role="menuitem"
          tabindex="-1"
        >
          <span
            class="ant-dropdown-menu-title-content"
          >
            Menu Item 1
          </span>
        </li>
        <li
          class="ant-dropdown-menu-item ant-dropdown-menu-item-only-child"
          data-menu-id="rc-menu-uuid-test-2"
          role="menuitem"
          tabindex="-1"
        >
          <span
            class="ant-dropdown-menu-title-content"
          >
            Menu Item 2
          </span>
        </li>
      </ul>
      <div
        aria-hidden="true"
        style="display: none;"
      />
    </div>
  </div>
</body>
`;

exports[`ConnectButton > Should show menu when hover button 1`] = `
<body>
  <div>
    <button
      class="ant-btn css-dev-only-do-not-override-1ck3jst ant-btn-default ant-dropdown-trigger css-dev-only-do-not-override-1ck3jst ant-web3-connect-button ant-dropdown-open"
      type="button"
    >
      <div
        class="ant-web3-connect-button-text css-dev-only-do-not-override-1ck3jst"
      >
        Connect Wallet
      </div>
    </button>
  </div>
  <div>
    <div
      class="ant-dropdown ant-slide-up-appear ant-slide-up-appear-prepare ant-slide-up css-dev-only-do-not-override-1ck3jst ant-dropdown-placement-bottomLeft"
      style="--arrow-x: 0px; --arrow-y: 0px; left: -1000vw; top: -1000vh; box-sizing: border-box;"
    >
      <ul
        class="ant-dropdown-menu ant-dropdown-menu-root ant-dropdown-menu-vertical ant-dropdown-menu-light css-dev-only-do-not-override-1ck3jst"
        data-menu-list="true"
        role="menu"
        tabindex="0"
      >
        <li
          class="ant-dropdown-menu-item"
          data-menu-id="rc-menu-uuid-test-copyAddress"
          role="menuitem"
          tabindex="-1"
        >
          <span
            aria-label="copy"
            class="anticon anticon-copy ant-dropdown-menu-item-icon"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="copy"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M832 64H296c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h496v688c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8V96c0-17.7-14.3-32-32-32zM704 192H192c-17.7 0-32 14.3-32 32v530.7c0 8.5 3.4 16.6 9.4 22.6l173.3 173.3c2.2 2.2 4.7 4 7.4 5.5v1.9h4.2c3.5 1.3 7.2 2 11 2H704c17.7 0 32-14.3 32-32V224c0-17.7-14.3-32-32-32zM350 856.2L263.9 770H350v86.2zM664 888H414V746c0-22.1-17.9-40-40-40H232V264h432v624z"
              />
            </svg>
          </span>
          <span
            class="ant-dropdown-menu-title-content"
          >
            Copy Address
          </span>
        </li>
        <li
          class="ant-dropdown-menu-item"
          data-menu-id="rc-menu-uuid-test-disconnect"
          role="menuitem"
          tabindex="-1"
        >
          <span
            aria-label="login"
            class="anticon anticon-login ant-dropdown-menu-item-icon"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="login"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <defs>
                <style />
              </defs>
              <path
                d="M521.7 82c-152.5-.4-286.7 78.5-363.4 197.7-3.4 5.3.4 12.3 6.7 12.3h70.3c4.8 0 9.3-2.1 12.3-5.8 7-8.5 14.5-16.7 22.4-24.5 32.6-32.5 70.5-58.1 112.7-75.9 43.6-18.4 90-27.8 137.9-27.8 47.9 0 94.3 9.3 137.9 27.8 42.2 17.8 80.1 43.4 112.7 75.9 32.6 32.5 58.1 70.4 76 112.5C865.7 417.8 875 464.1 875 512c0 47.9-9.4 94.2-27.8 137.8-17.8 42.1-43.4 80-76 112.5s-70.5 58.1-112.7 75.9A352.8 352.8 0 01520.6 866c-47.9 0-94.3-9.4-137.9-27.8A353.84 353.84 0 01270 762.3c-7.9-7.9-15.3-16.1-22.4-24.5-3-3.7-7.6-5.8-12.3-5.8H165c-6.3 0-10.2 7-6.7 12.3C234.9 863.2 368.5 942 520.6 942c236.2 0 428-190.1 430.4-425.6C953.4 277.1 761.3 82.6 521.7 82zM395.02 624v-76h-314c-4.4 0-8-3.6-8-8v-56c0-4.4 3.6-8 8-8h314v-76c0-6.7 7.8-10.5 13-6.3l141.9 112a8 8 0 010 12.6l-141.9 112c-5.2 4.1-13 .4-13-6.3z"
              />
            </svg>
          </span>
          <span
            class="ant-dropdown-menu-title-content"
          >
            Disconnect
          </span>
        </li>
      </ul>
      <div
        aria-hidden="true"
        style="display: none;"
      />
    </div>
  </div>
</body>
`;

exports[`ConnectButton > Should show menu when hover button 2`] = `
<body>
  <div>
    <button
      class="ant-btn css-dev-only-do-not-override-1ck3jst ant-btn-default ant-dropdown-trigger css-dev-only-do-not-override-1ck3jst ant-web3-connect-button ant-dropdown-open"
      type="button"
    >
      <div
        class="ant-web3-connect-button-text css-dev-only-do-not-override-1ck3jst"
      >
        <div
          class="ant-space css-dev-only-do-not-override-1ck3jst ant-space-horizontal ant-space-align-center ant-space-gap-row-small ant-space-gap-col-small ant-web3-address css-dev-only-do-not-override-1ck3jst"
        >
          <div
            class="ant-space-item"
          >
            <span>
              0x21CD...Fd3B
            </span>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div>
    <div
      class="ant-dropdown ant-slide-up-appear ant-slide-up-appear-start ant-slide-up css-dev-only-do-not-override-1ck3jst ant-dropdown-placement-bottomLeft"
      style="--arrow-x: 0px; --arrow-y: 0px; left: -1000vw; top: -1000vh; box-sizing: border-box; transition: none;"
    >
      <ul
        class="ant-dropdown-menu ant-dropdown-menu-root ant-dropdown-menu-vertical ant-dropdown-menu-light css-dev-only-do-not-override-1ck3jst"
        data-menu-list="true"
        role="menu"
        tabindex="0"
      >
        <li
          class="ant-dropdown-menu-item"
          data-menu-id="rc-menu-uuid-test-copyAddress"
          role="menuitem"
          tabindex="-1"
        >
          <span
            aria-label="copy"
            class="anticon anticon-copy ant-dropdown-menu-item-icon"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="copy"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M832 64H296c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h496v688c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8V96c0-17.7-14.3-32-32-32zM704 192H192c-17.7 0-32 14.3-32 32v530.7c0 8.5 3.4 16.6 9.4 22.6l173.3 173.3c2.2 2.2 4.7 4 7.4 5.5v1.9h4.2c3.5 1.3 7.2 2 11 2H704c17.7 0 32-14.3 32-32V224c0-17.7-14.3-32-32-32zM350 856.2L263.9 770H350v86.2zM664 888H414V746c0-22.1-17.9-40-40-40H232V264h432v624z"
              />
            </svg>
          </span>
          <span
            class="ant-dropdown-menu-title-content"
          >
            Copy Address
          </span>
        </li>
        <li
          class="ant-dropdown-menu-item"
          data-menu-id="rc-menu-uuid-test-disconnect"
          role="menuitem"
          tabindex="-1"
        >
          <span
            aria-label="login"
            class="anticon anticon-login ant-dropdown-menu-item-icon"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="login"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <defs>
                <style />
              </defs>
              <path
                d="M521.7 82c-152.5-.4-286.7 78.5-363.4 197.7-3.4 5.3.4 12.3 6.7 12.3h70.3c4.8 0 9.3-2.1 12.3-5.8 7-8.5 14.5-16.7 22.4-24.5 32.6-32.5 70.5-58.1 112.7-75.9 43.6-18.4 90-27.8 137.9-27.8 47.9 0 94.3 9.3 137.9 27.8 42.2 17.8 80.1 43.4 112.7 75.9 32.6 32.5 58.1 70.4 76 112.5C865.7 417.8 875 464.1 875 512c0 47.9-9.4 94.2-27.8 137.8-17.8 42.1-43.4 80-76 112.5s-70.5 58.1-112.7 75.9A352.8 352.8 0 01520.6 866c-47.9 0-94.3-9.4-137.9-27.8A353.84 353.84 0 01270 762.3c-7.9-7.9-15.3-16.1-22.4-24.5-3-3.7-7.6-5.8-12.3-5.8H165c-6.3 0-10.2 7-6.7 12.3C234.9 863.2 368.5 942 520.6 942c236.2 0 428-190.1 430.4-425.6C953.4 277.1 761.3 82.6 521.7 82zM395.02 624v-76h-314c-4.4 0-8-3.6-8-8v-56c0-4.4 3.6-8 8-8h314v-76c0-6.7 7.8-10.5 13-6.3l141.9 112a8 8 0 010 12.6l-141.9 112c-5.2 4.1-13 .4-13-6.3z"
              />
            </svg>
          </span>
          <span
            class="ant-dropdown-menu-title-content"
          >
            Disconnect
          </span>
        </li>
      </ul>
      <div
        aria-hidden="true"
        style="display: none;"
      />
    </div>
  </div>
</body>
`;
